src/main/java/com/cloudbees/jenkins/support/util/Anonymizer.java

```diff
@@ -0,0 +1,207 @@
+            }
+        }
+
+        if (settings.isAnonymizeUsers()) {
+            for (User user : User.getAll()) {
+                anonymizeName(user.getDisplayName(), "user", "", false);
+            }
+        }
+
+        save();
+    }
+
+    public static String anonymize(String input) {
+        String anonymized = input;
+        for (String key : ORDER) {
+            anonymized = anonymized.replaceAll("(?i)\\b(" + key + ")\\b", ANON_MAP.get(key));
+        }
+        return anonymized;
+    }
+
+    public static Map<String, String> getAnonymizedItems() {
+        return Collections.unmodifiableMap(ANON_MAP);
+    }
+
+    public static Map<String, String> getDisplayItems() {
+        if (LAST_REFRESH + DISPLAY_REFRESH_INTERVAL < System.currentTimeMillis()) {
+            refresh();
+        }
+        return DISPLAY;
+    }
+
+    private static String anonymizePath(String original, String prefix, boolean save) {
+        String[] labels = original.split("/");
+        StringBuilder oldPath = new StringBuilder();
+        String newPath = "";
+        for (int i = 0; i < labels.length; i++) {
+            oldPath.append(labels[i]);
+            newPath = anonymizeName(oldPath.toString(), prefix, newPath.toString(), save);
+            if (i != labels.length - 1) {
+                oldPath.append("/");
+                newPath += "/";
+            }
+        }
+        if (original.endsWith("/")) {
+            newPath += "/";
+        }
+        return newPath.toString();
+    }
+
+    private static String anonymizeName(String original, String prefix, String newPath, boolean save) {
+        if (SupportPlugin.getInstance().getExcludedWordsFromAnonymization().contains(original.toLowerCase(Locale.ENGLISH))) {
+            return original;
+        } else if (!ANON_MAP.containsKey(original)) {
+            ORDER.add(original);
+            String anonymized = newPath + prefix + "_" + GENERATOR.next();
+            ANON_MAP.put(original, anonymized);
+            ANON_MAP.put(Functions.escape(original), anonymized);
+            DISPLAY.put(original, anonymized);
+            for (String separator : SEPARATORS) {
+                String replaced = original.replaceAll("/", separator);
+                ANON_MAP.put(replaced, anonymized);
+                ANON_MAP.put(Functions.escape(replaced), anonymized);
+            }
+
+            if (save) {
+                save();
+            }
+        }
+        return ANON_MAP.get(original);
+    }
+
+    private static XmlFile getConfigXml() {
+        return new XmlFile(Jenkins.XSTREAM, new File(Jenkins.get().getRootDir(), "secrets/anonymized-names.xml"));
+    }
+
+    @SuppressWarnings("unchecked")
+    private static synchronized void load() {
+        XmlFile config = getConfigXml();
+        if (config.exists()) {
+            try {
+                ANON_MAP.clear();
+                ANON_MAP.putAll((Map<? extends String, ? extends String>) config.read());
```

Comment:
We need to also load the other fields, otherwise the UI and anonymization don't work after a restart because `DISPLAY` and `ORDER` stay empty. We could derive them from ANON_MAP like the following, but it might make more sense to store them independently so that original and escaped names don't end up in the display.

    DISPLAY.putAll(ANON_MAP);
    ORDER.addAll(ANON_MAP.keySet());